Enhance/correct some examples in expressions.md

RexJaeschke opened the pull request:
As a result of using the test harness, I made the following edits:

- used `System` namespace
- added a wrapper class [later, undid this]
- combined consecutive parts of code and text in a single part that would compile [later, undid this]

RexJaeschke committed:
Correct/reformat example

RexJaeschke committed:
Correct/reformat example

jskeet commented on standard/expressions.md:
```diff
@@ -71,11 +71,16 @@ Static binding takes place at compile-time, whereas dynamic binding takes place
-> Console.WriteLine(5); // static binding to Console.WriteLine(int)
-> Console.WriteLine(o); // static binding to Console.WriteLine(object)
-> Console.WriteLine(d); // dynamic binding to Console.WriteLine(int)
+> using System;
```
I'm not sure whether this change actually makes it more readable.

We can definitely create a template that has some common using directives and then puts the code from the example into a method in a library. I think I'd prefer that over adding the method declaration explicitly in this example, but I'm willing to be persuaded.

RexJaeschke replied:
removed using directive

jskeet commented on standard/expressions.md:
```diff
@@ -680,23 +685,25 @@ When arguments are omitted from a function member with corresponding optional pa
 > class Chooser
 > {
->     static Random rand = new Random();
+>     static System.Random rand = new System.Random();
```
I think it would be more idiomatic to have a `using System` directive.

RexJaeschke replied:
Examples as they existed *before* I proposed changes, are inconsistent; some have a `System.` prefix and some have a `using System;`.

Some observations:
1. Following our style, using the `using <namespace>;` approach adds 2 lines, that one plus a blank line.
2. In a few cases, we use a very long prefix, such as `System.Collections.Generic.`, which is a bit intrusive in code, and may be better suited as a `using <namespace>;`.

RexJaeschke replied:
removed name qualifiers

jskeet commented on standard/expressions.md:
```diff
@@ -888,22 +897,25 @@ The ***inferred return type*** is determined as follows:
+>     {
+>         return f2(f1(value));
+>     }
+>     static void M()
```
Add a blank line between methods.

jskeet commented on standard/expressions.md:
```diff
@@ -888,22 +897,25 @@ The ***inferred return type*** is determined as follows:
-> static Z F<X,Y,Z>(X value, Func<X,Y> f1, Func<Y,Z> f2)
+> using System;
+>
+> class A
```
Again, I suspect we could have a library template for "there's already a class declaration, put the given code inside it".

RexJaeschke replied:
I left it as I proposed. The example is still marked "incomplete," so we can revisit it later.

jskeet commented on standard/expressions.md:
```diff
@@ -1039,27 +1051,27 @@ Given two types `T₁` and `T₂`, `T₁` is a ***better conversion target*** th
->     int F1(int i);             // will pick non-generic
->     void F2(I1<U> a);          // Valid overload
->     void F2(I2<U> a);
+>     abstract public int F1(U u);               // Overload resulotion for G<int>.F1
```
Let's fix the typo for resolution at the same time :)

jskeet commented on standard/expressions.md:
```diff
@@ -1039,27 +1051,27 @@ Given two types `T₁` and `T₂`, `T₁` is a ***better conversion target*** th
->     void F2(I2<U> a);
+>     abstract public int F1(U u);               // Overload resulotion for G<int>.F1
+>     abstract public int F1(int i);             // will pick non-generic
+>     abstract public void F2(I1<U> a);          // Valid overload
```
I'd be tempted to add a blank line between each set of names (so both F1s together, then both F2s, and likewise for F3, F4, F5, F6 in G2.)

jskeet commented on standard/expressions.md:
```diff
@@ -2127,23 +2141,25 @@ A member initializer that specifies a collection initializer after the equals si
 When an initializer target refers to an indexer, the arguments to the indexer shall always be evaluated exactly once. Thus, even if the arguments end up never getting used (e.g., because of an empty nested initializer), they are evaluated for their side effects.
 
-> *Example*: The following class represents a point with two coordinates:
+> *Example*: The following shows a class that represents a point with two coordinates, and teh creation and initializatiob of an instance:
```
teh => the
initializiob => initialization

jskeet commented on standard/expressions.md:
```diff
@@ -2152,27 +2168,37 @@ When an initializer target refers to an indexer, the arguments to the indexer sh
-> where `__a` is an otherwise invisible and inaccessible temporary variable. The following class represents a rectangle created from two points:
+> where `__a` is an otherwise invisible and inaccessible temporary variable. 
+>
+> The following class ahows a rectangle created from two points, and the creation and initialization of a rectangle instance:
```
ahows => shows

jskeet replied:
And probably rectangle => `Rectangle` at the end of the sentence

jskeet commented on standard/expressions.md:
````diff
@@ -2152,27 +2168,37 @@ When an initializer target refers to an indexer, the arguments to the indexer sh
 >
-> ```csharp
-> Rectangle r = new Rectangle
+> class A
````
This is a trickier one in terms of avoiding declaring an extra class and method - but having *some* top-level code. Would like to discuss this in our next meeting.

RexJaeschke replied:
@jskeet Did we get any direction on how to handle this?

jskeet replied:
I would suggest creating additional files for Point and Rectangle, and specifying them as required.

RexJaeschke replied:
Done! I removed the redundant, hard-coded 1x`Point` and 2x`Rectangle` type declarations, replacing them with the following additions to the "additional-files" folder: Point.cs, Rectangle.cs, and RectangleWithConstructorPointAllocation.cs, and using the appropriate `additionalFiles:` in the examples' metadata.

jskeet commented on standard/expressions.md:
```diff
@@ -2189,27 +2215,34 @@ When an initializer target refers to an indexer, the arguments to the indexer sh
+>     {
+>         Rectangle r = new Rectangle
+>         {
+>             P1 = new Point { X = 0, Y = 1 },
```
You've changed the code so that it doesn't demonstrate what it's meant to... these should be

```csharp
P1 = { X = 0, Y = 1 },
P2 = { X = 2, Y = 3 }
```

jskeet commented on standard/expressions.md:
```diff
@@ -2189,27 +2215,34 @@ When an initializer target refers to an indexer, the arguments to the indexer sh
+> public class Rectangle
+> {
+>     public Point P1 { get; } = new Point();
+>     public Point P2 { get; } = new Point();}
```
Line break before the }

jskeet commented on standard/expressions.md:
```diff
@@ -3832,7 +3868,7 @@ Unless one of these conditions is true, a binding-time error occurs.
 >       if (x == null)
 >       {
->           throw new ArgumentNullException();
+>           throw new System.ArgumentNullException();
```
We should probably discuss what convention we want for explicit use of namespaces vs using directives.

RexJaeschke replied:
removed qualifier, and marked as "Incomplete"

RexJaeschke committed:
minor tweak

RexJaeschke committed:
fix errors in previous commit

jskeet reviewed (commented):
Thanks - let's leave the PR open until we've discussed it next week. (I'll have a whole agenda item about this...)

RexJaeschke committed:
make minor tweaks

RexJaeschke committed:
Merge branch 'dotnet:draft-v7' into RexJaeschke-patch-8

RexJaeschke committed:
add Point and Rectangle support files

I removed the redundant, hard-coded 1x`Point` and 2x`Rectangle` type declarations, replacing them with the following additions to the "additional-files" folder: Point.cs, Rectangle.cs, and RectangleWithConstructorPointAllocation.cs, and using the appropriate `additionalFiles:` in the examples' metadata.

RexJaeschke commented:
@jskeet I believe this is ready for your review.

RexJaeschke committed:
Create Point.cs

RexJaeschke committed:
Create Rectangle.cs

RexJaeschke committed:
Create RectangleWithConstructorPointAllocation.cs

jskeet commented on standard/expressions.md:
```diff
@@ -687,7 +690,7 @@ When arguments are omitted from a function member with corresponding optional pa
 When a generic method is called without specifying type arguments, a ***type inference*** process attempts to infer type arguments for the call. The presence of type inference allows a more convenient syntax to be used for calling a generic method, and allows the programmer to avoid specifying redundant type information.
 
-> *Example*: Given the method declaration:
+> *Example*: Given the method declaration and use, it is possible to invoke the `Choose` method without explicitly specifying a type argument:
```
Hmm... this doesn't quite read right to me, as "it is possible" suggests that it's possible to add more code to invoke the Choose method without explicitly specifying a type argument.

Is it worth just removing the sentence entirely, leaving just "Example:"? I can't remember whether every example has a description.

RexJaeschke replied:
OK; I removed the lead-in sentence.

jskeet commented on standard/expressions.md:
```diff
@@ -1059,27 +1070,31 @@ Given two types `T₁` and `T₂`, `T₁` is a ***better conversion target*** th
->     int F1(int i);             // will pick non-generic
->     void F2(I1<U> a);          // Valid overload
->     void F2(I2<U> a);
+>     abstract public int F1(U u);               // Overload resolution for G<int>.F1
```
I'd reorder all of these to "public abstract" - and maybe make the top-level class public as well?
(I do find all the "clutter" of public abstract to be unfortunate. I'm not sure what a better solution would be.)

RexJaeschke replied:
Done

jskeet commented on standard/expressions.md:
```diff
@@ -1059,27 +1070,31 @@ Given two types `T₁` and `T₂`, `T₁` is a ***better conversion target*** th
+>     abstract public void F5(U u1, I1<V> v2);   // Valid overload
+>     abstract public void F5(V v1, U u2);
+>
+>     abstract public void F6(ref U u);          // valid overload
```
```suggestion
>     abstract public void F6(ref U u);          // Valid overload
```

RexJaeschke replied:
Done

jskeet commented on standard/expressions.md:
```diff
@@ -2150,7 +2167,7 @@ A member initializer that specifies a collection initializer after the equals si
 When an initializer target refers to an indexer, the arguments to the indexer shall always be evaluated exactly once. Thus, even if the arguments end up never getting used (e.g., because of an empty nested initializer), they are evaluated for their side effects.
 
-> *Example*: The following class represents a point with two coordinates:
+> *Example*: The following shows a class that represents a point with two coordinates, and the creation and initialization of an instance:
```
Revert this change, given that the immediate example doesn't initialize an instance?

RexJaeschke replied:
Done

jskeet commented on standard/expressions.md:
````diff
@@ -2235,7 +2257,7 @@ When an initializer target refers to an indexer, the arguments to the indexer sh
 > ```
 >
-> which has the same effect as
+> This has the same effect as
````
Should this have metadata so that we compile it?

RexJaeschke replied:
Added metadata placeholder `<!-- Untested$Example: { ...`, so this will get processed in the next round.

jskeet commented on standard/expressions.md:
````diff
@@ -2287,33 +2309,37 @@ A collection initializer consists of a sequence of element initializers, enclose
 >
-> ```csharp
-> var contacts = new List<Contact>
+> class A
````
I wonder whether we should have a convention of `class Example` or similar?

RexJaeschke replied:
I don't have an opinion on this.

jskeet commented on standard/expressions.md:
````diff
@@ -2359,18 +2385,18 @@ Except in an unsafe context ([§22.2](unsafe-code.md#222-unsafe-contexts)), the
 <!-- IncompleteExample: {template:"standalone-console", name:"ArrayCreationExpressions1"} -->
 ```csharp
-new int[,] {{0, 1}, {2, 3}, {4, 5}}
+var a = new int[,] {{0, 1}, {2, 3}, {4, 5}}
````
I'd expect this and the example below to require semi-colons at the end.

jskeet replied:
(But then they could be marked as complete?)

RexJaeschke replied:
Done, and will "complete" them in the next round.

jskeet commented on standard/expressions.md:
```diff
@@ -4097,7 +4122,7 @@ Note that some conversions, such as user defined conversions, are not possible w
 >
 >     public T G<T>(object o)
->         where T : Attribute
+>         where T : System.Attribute
```
Do we need this change, if standalone-lib is going to have using directives?

RexJaeschke replied:
reversed.

jskeet commented on standard/expressions.md:
```diff
@@ -4692,16 +4720,29 @@ When not captured, there is no way to observe exactly how often a local variable
+>             result[i] = () => Console.WriteLine(x);
+>         }
+>         return result;
+> }
```
```suggestion
>     }
```

RexJaeschke replied:
Done

jskeet commented on standard/expressions.md:
```diff
@@ -4828,11 +4882,13 @@ The simplest form of an anonymous function is one that captures no outer variabl
     static void F()
     {
-        D d = () => Console.WriteLine("test");
+        D d = () => System.Console.WriteLine("test");
```
Again, revert this?

jskeet replied:
(And a few changes below, at lines 4911, 4928, 4950, 4992, 5014, 5055.)

RexJaeschke replied:
All reversed

jskeet commented on standard/expressions.md:
````diff
@@ -5715,6 +5785,7 @@ A generic type `C<T>` supports the query-expression-pattern if its public member
 
 <!-- IncompleteExample: {template:"standalone-lib", name:"Query-expressionPattern", expectedErrors:["x","x"], expectedWarnings:["x","x"]} -->
 ```csharp
+delegate R Func<T1,R>(T1 arg1);
````
Do we need either of these declarations now, if we've got using directives for System?

RexJaeschke replied:
I'll review this in the next round when I try to complete this example.

RexJaeschke committed:
Merge branch 'dotnet:draft-v7' into RexJaeschke-patch-8

RexJaeschke commented:
@jskeet I've fixed all the errors and deferred a couple of things to the next round, at which time I'll factor in the new templates and additional-files. Once this PR has been merged, I'll start that next round.

RexJaeschke committed:
Apply fixes requested by Jon

jskeet commented on standard/expressions.md:
```diff
@@ -72,11 +72,14 @@ Static binding takes place at compile-time, whereas dynamic binding takes place
-> Console.WriteLine(5); // static binding to Console.WriteLine(int)
-> Console.WriteLine(o); // static binding to Console.WriteLine(object)
-> Console.WriteLine(d); // dynamic binding to Console.WriteLine(int)
+> static void M()
```
Do we need this declaration? If we used `code-in-main` as the template, could we basically leave the example as it was?

RexJaeschke replied:
Agreed; reverted.

jskeet commented on standard/expressions.md:
````diff
@@ -859,6 +864,8 @@ The ***inferred return type*** is determined as follows:
 >
 > <!-- Example: {template:"standalone-lib", name:"InferredReturnType1"} -->
 > ```csharp
+> using System.Collections.Generic;
````
Is this now redundant?

RexJaeschke replied:
Agreed; removed

jskeet commented on standard/expressions.md:
````diff
@@ -2287,33 +2310,37 @@ A collection initializer consists of a sequence of element initializers, enclose
 >
 > <!-- Example: {template:"standalone-lib", name:"CollectionInitializers2"} -->
 > ```csharp
+> using System.Collections.Generic;
````
Remove unnecessary using directive?

RexJaeschke replied:
agreed; removed

jskeet commented on standard/expressions.md:
```diff
@@ -3888,7 +3914,7 @@ Unless one of these conditions is true, a binding-time error occurs.
 
 <!-- markdownlint-enable MD028 -->
-> *Example*: The following example checks whether an argument of an unconstrained type parameter type is `null`.
+> *IncompleteExample*: The following example checks whether an argument of an unconstrained type parameter type is `null`.
```
I don't think this was where you intended to insert `Incomplete`.

RexJaeschke replied:
agreed; removed

jskeet commented on standard/expressions.md:
````diff
@@ -4507,7 +4533,7 @@ Anonymous functions in an argument list participate in type inference and overlo
 > <!-- Example: {template:"standalone-lib", name:"OverloadResolution1"} -->
 > ```csharp
-> class ItemList<T> : List<T>
+> class ItemList<T> : System.Collections.Generic.List<T>
````
Do we need this qualification any more, with the addition of the using directives?

RexJaeschke replied:
agreed; removed

jskeet reviewed (commented):
Just a last few comments now.

RexJaeschke commented:
@jskeet ready for your review. I'm getting a markdownlinterror, but I don't see that as being relating to the md file content.

RexJaeschke committed:
Fix nits found by Jon

RexJaeschke committed:
Merge branch 'draft-v7' into RexJaeschke-patch-8

RexJaeschke commented:
@jskeet I think we're good to go with this.
